_posts/2021-11-18-keras-updates/keras-updates.Rmd

```diff
@@ -0,0 +1,159 @@
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+Before we even talk about new features, let us answer the obvious question. Yes, there will be a second edition of *Deep Learning for R*! Reflecting what has been going on in the meantime, the new edition covers an extended set of proven architectures; at the same time, you'll find that intermediate-to-advanced designs already present in the first edition have become rather more intuitive to implement, thanks to the new low-level enhancements alluded to in the summary.
+
+But don't get us wrong -- the scope of the book is completely unchanged. It still is the perfect choice for people new to machine learning and deep learning. Starting from the basic ideas, it systematically progresses to intermediate and advanced topics, leaving you with both conceptual understanding and a bag of useful application templates.
+
+*Now, what has been going on with Keras?*
+
+## State of the ecosystem
+
+Let us start with a characterization of the ecosystem, and a few words on history.
+
+In this post, when we say *Keras*, we mean R -- as opposed to Python -- *Keras*[^1]. Now, this immediately translates to the R package `keras`. But `keras` alone wouldn't get you far. While `keras` provides the high-level functionality -- neural network layers, optimizers, workflow management, and more -- the basic data structure operated upon, *tensors*, lives in `tensorflow`. Thirdly, as soon as you'll need to perform less-then-trivial pre-processing, or can no longer keep the whole training set in memory because of its size, you'll want to look into `tfdatasets`.
+
+[^1]: Originally, both Keras and Tensorflow are Python libraries, bound to by R at runtime. In practice, this means that those Python libraries have to be present on the R user's machine. This is in contrast to R `torch` , which has no Python dependencies. (Instead, it directly binds to a C++ library.)
+
+So it is these three packages -- `tensorflow`, `tfdatasets`, and `keras` -- that should be understood by "Keras" in the current context[^2]. (The R-Keras ecosystem, on the other hand, is quite a bit bigger. But other packages, such as `tfruns` or `cloudml`, are more decoupled from the core.)
+
+[^2]: Definitely, `tfautograph` belongs in this group as well, providing even-lower-level support for already-low-level `tensorflow`. But normally, the user will not have to call its functions directly.
+
+Matching their tight integration, the aforementioned packages tend to follow a common release cycle, itself dependent on the underlying Python library, [TensorFlow](http://tensorflow.org). For each of `tensorflow`, `tfdatasets`, and `keras` , the current CRAN version is 2.7.0, reflecting the corresponding Python version. The synchrony of versioning between the two Kerases, R and Python, seems to indicate that their fates had developed in similar ways. Nothing could be less true, and knowing this can be helpful.
```

Comment:
```suggestion
Matching their tight integration, the aforementioned packages tend to follow a common release cycle, itself dependent on the underlying Python library, [TensorFlow](http://tensorflow.org). For `tensorflow`, `tfdatasets`, and `keras` , the current CRAN version is 2.7.0, reflecting the corresponding Python version. The synchrony of versioning between the two Kerases, R and Python, seems to indicate that their fates had developed in similar ways. Nothing could be less true, and knowing this can be helpful.
```

Reply:
> The synchrony of versioning between the two Kerases, R and Python, seems to indicate that their fates had developed in similar ways. Nothing could be less true, and knowing this can be helpful.

I would pare this down and re-work it into the positive so that what you're saying is clearer (e.g. However, despite the synchrony of their versioning, the two Kerases, R and Python, have developed in different ways…) <- I don't actually like my wording, but the take-home message is that you don't get to the end of the sentence thinking the opposite of what is true (hopefully that makes sense)

Reply:
Honestly, i think that as the sentence is short, i could leave that part of rhetoric in ...